Idempotent is not the standard for 0-RTT (#4394)
quicwg/base-drafts: draft-ietf-quic-tls.md

means that restrictions on their use are necessary to avoid replay attacks on
the protocol.

Of the frames defined in {{QUIC-TRANSPORT}}, only the STREAM frame is

> potentially unsafe for use with 0-RTT as it carries application data.

Reviewer: Let me nitpick, as you seem to be wondering about frames that might have issues.

I think that this sentence is slightly incorrect. There are other frames that also carry application data. Specifically, CONNECTION_CLOSE (type 0x1d) and RESET_STREAM carry information defined by the application, and we do not require the integers carried by those frames to be replay-safe. See #3440.

I think we should either discuss about "application data" in general, or name the three frames (STREAM, RESET_STREAM, CONNECTION_CLOSE (type 0x1d)) that carry information defined by the application.

Author: Excellent point.  Will amend.

means that restrictions on their use are necessary to avoid replay attacks on
the protocol.

Of the frames defined in {{QUIC-TRANSPORT}}, the STREAM, RESET_STREAM, and

> CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they

Author: The error codes (and message for CONNECTION_CLOSE) carry application information that might be acted upon by an application.  It's a fine point, but worth covering.

> CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they

Reviewer 2: Fair point on the error codes, but I'm still having a very difficult time understanding the scenario of interest here.  RESET_STREAM has no impact unless you send STREAM frames, so let's focus on CONNECTION_CLOSE.

If I replay only a CONNECTION_CLOSE with no STREAM frames sent on the connection, what could that do?

> CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they

Reviewer 3: Imagine if your application protocol does something that's both important to the user and non-idempotent when the server receives an application layer connection close. Letting an attacker replay that would be bad. I think it's worth having some text for designers of application protocols that highlight this risk.

> CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they

Reviewer 2: The canonical example is a bank transaction where each request deposits a dollar in someone else's account.  Using a connection close error code to convey such a transaction seems implausible to me, so I'm not saying it's impossible, just that I can't construct an example that causes any issues.

> CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they

Reviewer 4: I don't know that there's an attack here, but RESET_STREAM can be sent without a STREAM, to retire that stream ID. I suppose if an application was applying semantic meaning to a particular stream-id, then canceling those stream-ids could be some sort of transaction? For example, I report a metric measurement on each stream, and if I reset that stream, then it's the default value? I dunno.

> CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they

Author: How about you have to pay a fine for every PROTOCOL_VIOLATION that clients report?

All of this is completely contrived, but as David says, better that you know.  I don't think that it needs a ton of explanation though.

> CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they can

Reviewer: ```suggestion
CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they
```

The three frames are defined to carry application data, I think.

Author: I was thinking about the transport-layer CONNECTION_CLOSE.  Maybe that doesn't matter.

carry application data.  Application data that is received in 0-RTT could cause
an application at the server to process the data multiple times rather than
just once. Additional actions taken by a server as a result of processing
replayed application data could have unwanted consequences. A client therefore
MUST only use 0-RTT for application data that is permitted by the application

> that is in use.

Reviewer 5: Do we want to have QUIC 0-RTT data and TLS 0-RTT data treated as equivalent for the purposes of being allowed by an application profile?  For example, the next paragraph cites RFC 8470 for HTTP 0-RTT data, but as written that document covers only TLS and not QUIC.
(AFAICT the security properties are the same so there shouldn't be a problem with letting "0-RTT" apply to both raw TLS and QUIC, but it seems like we should consciously choose to do that instead of implicitly doing it.)

Author: I think that it is OK that we treat HTTP/3 differently to HTTP/2, even if the underlying principles are the same.

It is relatively trivial for a protocol that is ported to QUIC to reuse guidance from TCP versions, as [HTTP/3 does in Section 10.9](https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-early-data).

Citing 8470 isn't perfect, as you say, because that means we are citing something that addresses TCP, so maybe I should acknowledge that directly.

An application protocol that uses QUIC MUST include a profile that defines
acceptable use of 0-RTT; otherwise, 0-RTT can only be used to carry QUIC frames
that do not carry application data. For example, a profile for HTTP is

> described in {{?HTTP-REPLAY=RFC8740}}.

Reviewer 6: Suggest rephrasing MUST since there is a way out in not using 0-RTT. For example, if an application is connecting over QUIC 0-RTT, the application protocol MUST ...

Author: I specifically chose this phrasing.  I know that prohibiting 0-RTT is a valid way to define a profile, but I think that is unnecessary for QUIC.  This would still allow that approach, but I would prefer a more nuanced profile that said "don't send application data in 0-RTT" than "don't send 0-RTT".

means that restrictions on their use are necessary to avoid replay attacks on
the protocol.

Of the frames defined in {{QUIC-TRANSPORT}}, the STREAM, RESET_STREAM, and
CONNECTION_CLOSE frames are potentially unsafe for use with 0-RTT as they
carry application data.  Application data that is received in 0-RTT could cause
an application at the server to process the data multiple times rather than
just once. Additional actions taken by a server as a result of processing
replayed application data could have unwanted consequences. A client therefore

> MUST only use 0-RTT for application data that is permitted by the application
> that is in use.

Reviewer 7: ```suggestion
MUST NOT use 0-RTT for application data unless specifically
requested by the application that is in use.
```

Aligning text with E.5 of RFC 8446. See https://github.com/quicwg/base-drafts/issues/4393#issuecomment-733634021
